### ISSUE TITLE
Клиент своей игры // изменение способа распространения вопросов (пре-кеширование)

### ISSUE THREAD

**Issue by BggClr** [VladimirKhil/SI#27]
На данный момент клиент получает вопрос и его медиа-данные в реальном времени в момент выбора игроком вопроса.
Данный способ вызывает проблемы с отображениям медиа-данных у разных игроков (в зависимости от нагрузки серверов СИ, расположения игроков, а также канала до серверов СИ), что ведет либо к преимуществу одних игроков над другими, либо к невозможности отыграть пак с картинками, музыкой итд

В идеале клиент своей игры должен получать весь пак перед игрой, по следующей схеме:

1. Хост при создании игры на каждый вопрос генерирует случайный ключ шифрования и запоминает его
# Хост конвертирует исходный пак в немного другой, шифруя каждый вопрос своим собственным ключом
# Хост отправляет

**Comment by VladimirKhil** [comment]
Я вот не уверен, что в локальном хранилище поместится весь пакет. А его ещё раззиповать надо будет. И как-то ресурсы на вход проигрывателю подать.
Возможно, вариант с повышением мозности сервера будет лучше.

**Comment by BggClr** [comment]
С раззиповыванием и передачей ресурсов проигрывателю в целом проблем нет - популярные плееры обычно из base64 умеют читать поток нормально, а вот с localstorage я немного погорячился конечно - туда даже раунд влезет вряд ли.

Повышение производительности сервера не факт что поможет, потому что проблема находится частично в скорости закачки от сервера до игроков. Как вариант можно попробовать подключить какой-нибудь кеширующий сервис для медиа типа cloudflare, к нему по идее у людей доступ гораздо быстрее , плюс нагрузку до сервака понизить может. Можно даже попробовать скармливать ему эти медиа-данные чуть заранее (от хоста), чтобы игроки уже прямо с него данные себе тянули

**Comment by VladimirKhil** [comment]
> С раззиповыванием и передачей ресурсов проигрывателю в целом проблем нет - популярные плееры обычно из base64 умеют читать поток нормально, а вот с localstorage я немного погорячился конечно - туда даже раунд влезет вряд ли.
> 
> Повышение производительности сервера не факт что поможет, потому что проблема находится частично в скорости закачки от сервера до игроков. Как вариант можно попробовать подключить какой-нибудь кеширующий сервис для медиа типа cloudflare, к нему по идее у людей доступ гораздо быстрее , плюс нагрузку до сервака понизить может. Можно даже попробовать скармливать ему эти медиа-данные чуть заранее (от хоста), чтобы игроки уже прямо с него данные себе тянули

Идея с Cloudflare хорошая, можно попробовать (если снова не нарваться на заблокированные Роскомнадзором IPники :))

**Comment by dimavoynov** [comment]
Привествую энтузиастов, давайте создадим группу в телеграмме. Адрес @si-teleg, заходите товарищи. Владимир, добавьте в инфу о группе.

**Comment by KazankovMarch** [comment]
Могу предложить на обсуждение альтернативное решение:
Клиент SIGame показывает контент не сразу после скачивания. Вместо этого он посылает сигнал серверу, что успешно получил контент и ждет отмашки от сервера, что можно показывать.
Сервер, в свою очередь, ждет, пока от всех клиентов не придет сигнал об успешном скачивании (возможно, ретраит закачку, если от клиент долго не может скачать контент). И как только получает сигнал от всех клиентов, то дает отмашку, что можно показывать контент. Таким образом показ контента начнется у всех одновременно

**Comment by CrafterKolyan** [comment]
Отмечу, что в предложении от @KazankovMarch есть небольшая проблема безопасности. Формально скачанные данные уже находятся у игрока и на самом деле он может получить к ним доступ. Таким образом в будущем возможно будет создан чит, позволяющий получить данное преимущество.
Тем не менее данное решение намного лучше, чем существующая в данный момент ситуация (когда можно сказать, что все пользуются читом). Также в пользу этого решения хочу сказать, что оно намного проще реализуется, нежели более честное решение с шифрованием. Тем не менее в некоторых наших играх мы замечали ситуацию, когда игрок вообще не мог получить ресурсы с сервера (то есть картинки не прогружались в течение очень долгого времени (5 минут)). В такой ситуации игра встанет у всех и надо дать возможность ведущему по кнопке Далее пропустить вопрос для всех.

Также если всё-таки будет выложен исходный код SIGame, я может и сам реализую решение данной проблемы.

**Comment by KazankovMarch** [comment]
Да, действительно есть такая проблема. Можно объединить идеи: сервер посылает контент зашифрованным, а во время "отмашки" посылает ключи клиентам

PS: проблема очень насущная,  мы с друзьями сейчас просто используем паки без медиа-контента (которые еще и сложно найти), иначе этот баг очень портит впечатление от игры. Мне кажется лучше будет сначала сделать быстрое решение, временно забыв про эту проблему с потенциальным читом. А шифрование или что-то другое добавить в рамках отдельной ишью/фичи

**Comment by CrafterKolyan** [comment]
Согласен. Я эту идею придумал уже давным давно, и описал в `README.md` своего Fork'а (https://github.com/CrafterKolyan/SI-Remastered), но есть проблема, состоящая в том, что мне придётся потратить на реализацию слишком много времени, т.к. я должен практически с нуля написать клиент игры SIGame и я понимаю, что в такой ситуации писать его на C# не очень хочется, т.к. я скорее всего очень быстро потеряю кроссплатформенность клиента, а проще сразу всё переписывать на Java. Тем не менее, сравнить код с чем-то реальным невозможно, документации по общению с сервером тоже нет. Поэтому одной из основных проблем является отсутствие полного кода SIGame, причём насколько я помню обсуждение о том, что код SIGame надо открыть было и @VladimirKhil сказал, что выложит его после некоторых "приукрашиваний", но пока новостей об этом нет. Если честно, мне будет достаточно любого кода, если надо будет сам отрефакторю и напишу все необходимые фишки, связанные с честностью игры и стабильностью работы клиента.

**Comment by KazankovMarch** [comment]
Блин, если вопрос стоит таким боком и в ближайшем будущем не появятся исходники, стоит подумать о реализации тонкой прокси, которая реализует этот функционал и кидает на вход "настоящему" клиенту медиа-контент, когда сервер кинет отмашку. Исходники сервера же полностью выложены?

**Comment by CrafterKolyan** [comment]
Есть вот такая часть серверов https://github.com/VladimirKhil/SI/blob/master/src/SICore/SICore.Network/Servers/ . Из полезного на первый взгляд там есть файл `Server.cs`. Насколько это то, что необходимо - не ясно. Конкретного проекта с сервером я так и не нашёл. Соответственно, мы не можем просто склонировать репозиторий, скомпилировать и запустить свой сервер из под коробки и не можем как-то просто проверять свои изменения. Идея с проксой с одной стороны может сработать, но это явно грязь, потому что мы не можем адекватно изменить исходник SIGame, а значит нам надо перехватывать все пакеты от приложения и самим их обрабатывать. Насколько это реально я не знаю. Также мне кажется, что это решение явно будет не очень стабильным. В такой ситуации я бы предпочёл переписать весь SIGame с нуля, чем делать такую обёртку. Единственное меня волнует совместимость с оригинальным сервером - хочется, чтобы она была. Совместимость с .siq пакетами - вообще не проблема, причём можно хоть все 4 версии пакетов уметь обрабатывать. Ещё есть проблема, что нужен хостинг для сервера, который жрёт ресурсы разработчиков, но при этом сам проект кажется не монетизируем.

**Comment by KazankovMarch** [comment]
Можно и без хостинга. Мы с друзьями всегда локалку поднимаем и кто-то хостит сервер на своей машине. С остальным согласен

**Comment by VladimirKhil** [comment]
- сервер пока выладывать не планируется. Но для сценариев прямого подключения, которые, как я понял, здесь являются приоритетными, он и не нужен;
- вся концепция с предзагрузкой вопросов действительно зависит от честности клиента (чего не хотелось бы) и совершенно непонятно как должна работать в веб-клиенте. А основной упор в развитии будет делаться именно на веб-версию;
- исходники остальной части десктопного клиента будут выложены после того, как сервер будет полностью защищён от атак хакнутым клиентом. Надеюсь, что это произойдёт в течение месяца.

**Comment by GooG2e** [comment]
Для прямого подключения важно реализовать нормально серверную часть клиента. Сейчас это огромная проблема, потому что пак может рандомно не раздаваться, кому-то раздаться, а кому-то нет.
Даже прослойка в виде кеширующего прокси-сервера не помогает спасти ситуацию т.к. иногда клиент даже один раз не может отдать видео файл.
Соответственно в данный момент интересует модификация как раз той части клиента, которая отвечает за раздачу контента при игре по прямому подключению.

**Comment by iPilot** [comment]
@BggClr серверу не обязательно после этого хранить медийку в памяти, достаточно оставить мапу "хэш/guid - ключ".
Контент шифруется при старте раунда, ключ шифрования и id контента отсылаются при выборе вопроса. Ожидается декрипт немного и показывается вопрос.
Синхронизация конечно хороший вариант, но нужно не забыть про таймаут клиента, а в данном случае он должен быть совсем небольшим (1 секунда?).

**Comment by VladimirKhil** [comment]
В режиме игры без фальстартов не получится применить, т.к. клиент будет иметь у себя готовый файл и не обязательно прервёт его воспроизведение при нажатии кем-то кнопки.

**Comment by VladimirKhil** [comment]
Исходники игры полностью выложены. Если есть желание реализовать один из предложенных вариантов - милости прошу!
